R2RDump fails parsing GcInfo when emitting PDB for System.Private.CoreLib compiled with Crossgen2 in release mode for ARM

OS: Windows
Architecture: ARM
Configuration: Release

Until I recently modified R2RDump to load GcInfo lazily, it was doing that when parsing the runtime function table (i.e. early in the R2R loading process). Because of this we observed the following GcInfo crash when emitting PDB symbol information, e.g. in this example run:

https://dev.azure.com/dnceng/public/_build/results?buildId=875055&view=logs&jobId=29c18220-1650-5afc-169a-2e7f66607d7e&j=29c18220-1650-5afc-169a-2e7f66607d7e&t=91035a01-1a51-53e6-8daf-e4f287926386

<pre>
  Filename: F:\workspace\_work\1\s\artifacts\bin\coreclr\windows.arm.Release\System.Private.CoreLib.dll
  OS: Windows
  Machine: ArmThumb2
  ImageBase: 0x00400000
  
EXEC : error : System.IndexOutOfRangeException: Index was outside the bounds of the array. [F:\workspace\_work\1\s\src\coreclr\crossgen-corelib.proj]
##[error]EXEC(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) System.IndexOutOfRangeException: Index was outside the bounds of the array.
     at ILCompiler.Reflection.ReadyToRun.NativeReader.ReadBits(Byte[] image, Int32 numBits, Int32& bitOffset) in /_/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/NativeReader.cs:line 110
     at ILCompiler.Reflection.ReadyToRun.Amd64.GcInfo.GetTransitions(Byte[] image, Int32& bitOffset) in /_/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/Amd64/GcInfo.cs:line 531
     at ILCompiler.Reflection.ReadyToRun.Amd64.GcInfo..ctor(Byte[] image, Int32 offset, Machine machine, UInt16 majorVersion) in /_/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/Amd64/GcInfo.cs:line 199
     at ILCompiler.Reflection.ReadyToRun.ReadyToRunMethod.ParseRuntimeFunctions() in /_/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunMethod.cs:line 504
     at ILCompiler.Reflection.ReadyToRun.ReadyToRunMethod.EnsureRuntimeFunctions() in /_/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/ReadyToRunMethod.cs:line 267
     at R2RDump.R2RDump.ProducePdbWriterMethods(ReadyToRunReader r2r)+MoveNext() in /_/src/coreclr/src/tools/r2rdump/R2RDump.cs:line 430
     at ILCompiler.PdbWriter.PdbWriter.WritePDBDataHelper(String dllPath, IEnumerable`1 methods) in /_/src/coreclr/src/tools/r2rdump/PdbWriter.cs:line 265
     at ILCompiler.PdbWriter.PdbWriter.WritePDBData(String dllPath, IEnumerable`1 methods) in /_/src/coreclr/src/tools/r2rdump/PdbWriter.cs:line 152
     at R2RDump.R2RDump.Dump(ReadyToRunReader r2r) in /_/src/coreclr/src/tools/r2rdump/R2RDump.cs:line 412
     at R2RDump.R2RDump.Run() in /_/src/coreclr/src/tools/r2rdump/R2RDump.cs:line 594
</pre>

/cc @dotnet/crossgen-contrib

--- a/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/Amd64/GcInfo.cs
+++ b/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/Amd64/GcInfo.cs
@@ -199,8 +199,8 @@ namespace ILCompiler.Reflection.ReadyToRun.Amd64
                 }
             }
 
-            int endByteOffset = (bitOffset + 7) >> 3;
-            Size = endByteOffset - offset;
+            int nextByteOffset = (bitOffset + 7) >> 3;
+            Size = nextByteOffset - offset;
         }
 
         public override string ToString()

--- a/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/Amd64/GcInfo.cs
+++ b/src/coreclr/src/tools/aot/ILCompiler.Reflection.ReadyToRun/Amd64/GcInfo.cs
@@ -103,7 +103,6 @@ namespace ILCompiler.Reflection.ReadyToRun.Amd64
 
             Version = ReadyToRunVersionToGcInfoVersion(majorVersion);
             int bitOffset = offset * 8;
-            int startBitOffset = bitOffset;
 
             ParseHeaderFlags(image, ref bitOffset);
 
@@ -185,17 +184,23 @@ namespace ILCompiler.Reflection.ReadyToRun.Amd64
 
             SlotTable = new GcSlotTable(image, machine, _gcInfoTypes, ref bitOffset);
 
-            // Try partially interruptible first
-            if (NumSafePoints > 0)
+            if (SlotTable.NumSlots > 0)
             {
-                LiveSlotsAtSafepoints = GetLiveSlotsAtSafepoints(image, ref bitOffset);
-            }
-            else
-            {
-                Transitions = GetTransitions(image, ref bitOffset);
+                if (NumSafePoints > 0)
+                {
+                    // Partially interruptible code
+                    LiveSlotsAtSafepoints = GetLiveSlotsAtSafepoints(image, ref bitOffset);
+                }
+                else
+                {
+                    // Fully interruptible code
+                    Debug.Assert(NumInterruptibleRanges > 0);
+                    Transitions = GetTransitions(image, ref bitOffset);
+                }
             }
 
-            Size = bitOffset - startBitOffset;
+            int endByteOffset = (bitOffset + 7) >> 3;
+            Size = endByteOffset - offset;
         }
 
         public override string ToString()
@@ -482,7 +487,7 @@ namespace ILCompiler.Reflection.ReadyToRun.Amd64
         /// <summary>
         /// based on end of <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/gcinfodecoder.cpp">GcInfoDecoder::EnumerateLiveSlots and GcInfoEncoder::Build</a>
         /// </summary>
-        public Dictionary<int, List<BaseGcTransition>> GetTransitions(byte[] image, ref int bitOffset)
+        private Dictionary<int, List<BaseGcTransition>> GetTransitions(byte[] image, ref int bitOffset)
         {
             int totalInterruptibleLength = 0;
             if (NumInterruptibleRanges == 0)
@@ -511,8 +516,8 @@ namespace ILCompiler.Reflection.ReadyToRun.Amd64
                 chunkPointers[i] = NativeReader.ReadBits(image, numBitsPerPointer, ref bitOffset);
             }
 
-            // Offset to m_Info2 containing all the info on register liveness
-            int info2Offset = (int)Math.Ceiling(bitOffset / 8.0) * 8;
+            // Offset to m_Info2 containing all the info on register liveness, which starts at the next byte
+            int info2Offset = (bitOffset + 7) & ~7;
 
             List<GcTransition> transitions = new List<GcTransition>();
             bool[] liveAtEnd = new bool[SlotTable.NumTracked]; // true if slot is live at the end of the chunk

--- a/src/coreclr/src/tools/r2rdump/TextDumper.cs
+++ b/src/coreclr/src/tools/r2rdump/TextDumper.cs
@@ -161,6 +161,7 @@ namespace R2RDump
 
                 if (_options.Raw)
                 {
+                    // TODO: Output RVAs for consistency with other DumpBytes calls
                     DumpBytes(gcInfo.Offset, (uint)gcInfo.Size, "", false);
                 }
             }

